gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/CallGraphBugChecker.java

```diff
@@ -0,0 +1,136 @@
+         * The edge does not contain any transitive calls from f to g (e.g. f calls h calls g)
+         * For more details, see {@code CallGraphBugCheckerTest} for the graph's specification
+         */
+        @VisibleForTesting // Ideally MethodCallGraph is its own class, but SuppressibleTreePathScanner is protected
+        protected final MutableValueGraph<MethodSymbol, Set<MethodInvocationTree>> callGraph =
+                ValueGraphBuilder.directed().allowsSelfLoops(true).build();
+
+        /**
+         * This exists to be passed to {@code EdgeConsumer}s.
+         *  e.g. if the CompilationUnit contains the chained call getProject().getLogger(), this will map
+         *      getProject() to getProject().getLogger()
+         *  In other words, it maps child calls to parent calls in the AST of the Compilation Unit.
+         */
+        private final Map<MethodInvocationTree, MethodInvocationTree> callToChainedCall = new HashMap<>();
+
+        public MethodCallGraph(VisitorState state) {
+            new SuppressibleTreePathScanner<Void, Optional<MethodSymbol>>(state) {
+                @Override
+                public Void visitMethod(MethodTree node, Optional<MethodSymbol> caller) {
+                    MethodSymbol methodSymbol = ASTHelpers.getSymbol(node);
+                    callGraph.addNode(methodSymbol);
+                    return super.visitMethod(node, Optional.of(methodSymbol));
+                }
+
+                @Override
+                public Void visitMethodInvocation(MethodInvocationTree node, Optional<MethodSymbol> caller) {
+                    MethodSymbol calleeSymbol = (MethodSymbol) ASTHelpers.getSymbol(node.getMethodSelect());
+                    callGraph.addNode(calleeSymbol);
+
+                    if (caller.isPresent()) {
+                        MethodSymbol callerSymbol = caller.get();
+                        Set<MethodInvocationTree> existingInvocations =
+                                callGraph.edgeValue(callerSymbol, calleeSymbol).orElseGet(HashSet::new);
+                        existingInvocations.add(node);
+                        callGraph.putEdgeValue(callerSymbol, calleeSymbol, existingInvocations);
+                    }
+
+                    return super.visitMethodInvocation(node, caller);
+                }
+            }.scan(state.getPath().getCompilationUnit(), Optional.empty());
+
+            new SuppressibleTreePathScanner<Void, Optional<MethodInvocationTree>>(state) {
+                @Override
+                public Void visitMethodInvocation(
+                        MethodInvocationTree call, Optional<MethodInvocationTree> chainedCall) {
+                    chainedCall.ifPresent(chained -> callToChainedCall.put(call, chained));
+                    return super.visitMethodInvocation(call, Optional.of(call));
+                }
+            }.scan(state.getPath().getCompilationUnit(), Optional.empty());
```

Comment:
surely you can combine this with the above scan? seems somewhat wasteful to process the same source file *twice*